apps/rule-manager/client/src/ts/components/hooks/useTags.ts

```diff
@@ -35,9 +49,110 @@ export function useTags() {
     setIsLoading(false);
   }
 
+  const fetchTagRuleCounts = async (): Promise<void> => {
+    setIsLoadingTagRuleCounts(true);
+
+    try {
+      const response = await fetch(`${location.origin}/api/tags/ruleCount`);
+      if (!response.ok) {
+        throw new Error(`Failed to fetch rules: ${response.status} ${response.statusText}`);
+      }
+      const tagsRuleCounts: TagRuleCounts = await response.json();
+      setTagRuleCounts(tagsRuleCounts);
+    } catch (error) {
+      setError(errorToString(error));
+    }
+    setIsLoadingTagRuleCounts(false);
+  }
+
+  const updateTag = async (tagForm: Tag) => {
+    setIsLoading(true);
+
+    // We would always expect the ruleForm to include an ID when updating a rule
+    if (!tagForm.id) return {status: 'error', errorMessage: "Update endpoint requires a tag ID"} as ErrorIResponse;
+
+    try {
+      const response = await fetch(`${location.origin}/api/tags/${tagForm.id}`, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(tagForm)
+      })
+
+      const parsedResponse = await responseHandler<Tag>(response);
+      if (parsedResponse.status === "ok") {
+        const updatedTag = parsedResponse.data;
+        setTags({...tags, [`${updatedTag.id}`]: updatedTag})
+      } else {
+        setError(parsedResponse.errorMessage);
+      }
+    } catch (error) {
+      setError(errorToString(error));
+    } finally {
+      setIsLoading(false);
+    }
+  }
+
+  const createTag = async (tagName: string) => {
```

Comment:
Doesn't necessarily have to be part of this PR, but I wonder if we might want to prevent the user from creating duplicate tags? With an error message that warns if the tag already exists. And perhaps the ability to search tags on the Tags page, so the user can check if a tag exists before trying to create it.

Reply:
These definitely sound like sensible bits of functionality - I'd be inclined to card them up as follow-up work as this PR is already fairly chunky.

Reply:
Duplicate tags is a good shout for a follow-up PR! Searching tags I think can come as and when we need it – while our tags exist on a single page I think we can probably go without, but as the list of tags grows this might become harder to avoid.